### DISCUSSION TITLE
MUI - one button - two different jumps (destination pages)?

### DISCUSSION THREAD

**Discussion by ViacheslavK9** [olikraus/u8g2#2351] [Q&A] [answered]
Hello, Oliver!
Forgot to say, happy new year!

i know it's too many questions, but maybe your tech will bring much more with this info for others.

So, my question is - do you think it's possible to have one button on a form with different assigned form depending on a condition?
I am looking onto that because don't want to duplicate form,
which is accessible from the two different branches on a menu tree.

Thanks!

**Comment by olikraus** [comment]
> do you think it's possible to have one button on a form with different assigned form depending on a condition?

Sure, why not. The key-event mapping is anyways made by the mui use, so it would be easy to add any condition.

(created issue: https://github.com/olikraus/u8g2/issues/2356)

**Reply by ViacheslavK9** [reply]
i guess construction like     

`MUIF_BUTTON("TR", mui_u8g2_btn_goto_wm_fi),
MUI_XYAT("TR", 25, 56, 52, " Do test  ")`

didnt help me, because we directly writing form number in a definition of the object.
Did not find any example on something like "OnPushButton" trigger behavior, 
maybe you can help? Tx..

**Reply by olikraus** [reply]
Each MUI program needs a button event procedure like this:
https://github.com/olikraus/u8g2/blob/c4f9cd9f8717661c46be16bfbcb0017d785db3c1/sys/arduino/u8g2_page_buffer/MUIInput2BtnBounce2/MUIInput2BtnBounce2.ino#L516-L526

My idea was to extend this procedure with something like this:
``` C++
   if ( selectBtn.pressed() ) { 
     if ( isSpecialActionRequired ) {
        // user has pressed select, do something special
     } else {
       mui.sendSelect(); // normal MUI behavior
       is_redraw = 1; 
     }
   }  
```

Note, that also some button libs support long button press, which might be useful, if you have only one button.

For the above Bounce2 lib, there is at least a previousDuration()/currentDuration() which might allow you to check how long the button was pressed (https://github.com/thomasfredericks/Bounce2). But I am not an expert on Bounce2... you need to figure out by yourself if this might help, but an example for a one button code could look like this:

``` C++
   if ( selectBtn.pressed() ) { 
     if ( selectBtn.previousDuration() > 300 ) { // long press
       mui.sendSelect(); 
       is_redraw = 1; 
     } else {
       mui.nextField(); 
       is_redraw = 1; 
     }
   }  
```
... not tested

**Reply by olikraus** [reply]
And of course, instead of "mui.sendSelect()" you could also jump to a specific menu.
So for example, if the user presses this long button press, then use "mui.gotoForm()"  (https://github.com/olikraus/u8g2/wiki/muiref#gotoform) to directly jump to a specific form.

**Reply by olikraus** [reply, marked as answer]
I have created and tested a MUI menu with only one button and Bounce2 lib: https://github.com/olikraus/u8g2/blob/master/sys/arduino/u8g2_page_buffer/MUIInput1BtnBounce2/MUIInput1BtnBounce2.ino

There was a little change required: Instead of Bounce2.pressed() function we must use the Bounce2.released() function.

But all in all it works nicely with only one button:
https://github.com/olikraus/u8g2/blob/8430f65c9e2bc7ca393bf0e9d8cf58a84ae081d7/sys/arduino/u8g2_page_buffer/MUIInput1BtnBounce2/MUIInput1BtnBounce2.ino#L513-L522


https://github.com/olikraus/u8g2/issues/2357

**Reply by ViacheslavK9** [reply]
Thank you very much, Oliver! My Project is on pause for now, but your answers were very helpfull. 👍
